pkg/operator/vspherecontroller/vspherecontroller.go

```diff
@@ -456,32 +462,35 @@ func (c *VSphereController) createVCenterConnection(ctx context.Context, infra *
 	if !ok {
 		return fmt.Errorf("cloud config %s/%s does not contain key %q", cloudConfigNamespace, cloudConfig.Name, cloudConfig.Key)
 	}
-	cfg := new(vsphere.VSphereConfig)
-	err = gcfg.ReadStringInto(cfg, cfgString)
+
+	config := vclib.VSphereConfig{}
+	err = config.LoadConfig(cfgString)
 	if err != nil {
 		return err
 	}
 
 	// We no longer use the ConfigMap to store the vSphere config, so make sure to delete it
 	c.deleteConfigMapIfExists(ctx, legacyConfigMapName, c.targetNamespace)
 
-	secret, err := c.secretLister.Secrets(c.targetNamespace).Get(cloudCredSecretName)
-	if err != nil {
-		return err
-	}
-	userKey := cfg.Workspace.VCenterIP + "." + "username"
-	username, ok := secret.Data[userKey]
-	if !ok {
-		return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, userKey)
-	}
-	passwordKey := cfg.Workspace.VCenterIP + "." + "password"
-	password, ok := secret.Data[passwordKey]
-	if !ok {
-		return fmt.Errorf("error parsing secret %q: key %q not found", cloudCredSecretName, passwordKey)
-	}
+	for _, vcenter := range infra.Spec.PlatformSpec.VSphere.VCenters {
```

Comment:
Will - `infra.Spec.PlatformSpec.VSphere.VCenters` populated for clusters upgraded from older versions of OCP?

Reply:
Good question.  I think this method is not pre-filtered to upgrade to FD infra.  I can add that change in to prevent backward compat issues.

Reply:
I have now pre-filtered the FD data.  The only down side I am thinking for these changes is where logic expecting no FDs to be defined will no longer get executed.